### ISSUE TITLE
Strange issue "RangeError: out of range index"

### ISSUE THREAD

**Issue by ericvolp12** [sidorares/node-mysql2#254]
I've written a query for a database that I've run 1,000 times in mysql workbench and it works flawlessly

Here is the query:
```var message_query = "SELECT `from`, `timestamp`, `body`, `user_id`, `message_id` FROM `messages` WHERE `user_id` = :user_id AND `from` = :from AND `deleted` = 0 LIMIT :num;"; ```

I'm executing the query as such, using an instance of a database accessing class that works with every other query I've used.

```db.execute(message_query, {user_id:user_id, from:senders[j], num:num}, function(rows){```

I've evaluated all of those variables right before the statement is executed, and they evaluate to the right user id, phone number, and number of responses.

When the command executes through the node-mysql2 function, my error log throws this strange error:

```RangeError: out of range index
    at RangeError (native)
    at Packet.readLengthCodedString (/home/user/ericv-phone/node_modules/mysql2/lib/packets/packet.js:271:22)
    at new BinaryRow (evalmachine.<anonymous>:18:27)
    at Execute.Query.row (/home/user/ericv-phone/node_modules/mysql2/lib/commands/query.js:196:13)
    at Execute.Command.execute (/home/user/ericv-phone/node_modules/mysql2/lib/commands/command.js:34:20)
    at PoolConnection.Connection.handlePacket (/home/user/ericv-phone/node_modules/mysql2/lib/connection.js:310:28)
    at PacketParser.onPacket (/home/user/ericv-phone/node_modules/mysql2/lib/connection.js:83:65)
    at PacketParser.executeStart (/home/user/ericv-phone/node_modules/mysql2/lib/packet_parser.js:39:12)
    at Socket.<anonymous> (/home/user/ericv-phone/node_modules/mysql2/lib/connection.js:95:31)
    at emitOne (events.js:77:13)```

I've spent 4 hours trying to solve this on my own but to no avail.

My execution looks like this:
```
connection.execute(query, data, function(err, rows){
            if(err){
                console.log(err);
                connection.release();
                callback(err);
            }else{
                connection.release();
                callback(rows);
            }
        });
```

Where data is the object I showed earlier, ```{user_id:user_id, from:senders[j], num:num}```

I've even tried hardcoding the values of user_id, from, and num in the query, without requiring any placeholders in the query, but no matter what, the query fails to execute without logging an error in my log, but it does throw an error in the exception log.

Can someone please tell me what the hell is going on?

Thanks!

**Comment by sidorares** [comment]
hi! which module version are you using? is that one from npm?

Can you make simplest possible example to reproduce? ( with schema creation and data )?

**Comment by sidorares** [comment]
Also can you post output with connection flag `debug: true`

**Comment by sidorares** [comment]
and maybe put in lib/packets/packet.js:272 `console.log(len, this.offset, this.buffer)` and post output before error

**Comment by ericvolp12** [comment]
Hey, so I'm using version "^0.15.8" according to my package.json.

I'll show a schema below.

```
CREATE TABLE `messages` (
  `message_id` int(11) NOT NULL AUTO_INCREMENT,
  `user_id` int(11) NOT NULL,
  `to` varchar(15) NOT NULL,
  `from` varchar(15) NOT NULL,
  `body` varchar(250) NOT NULL,
  `timestamp` timestamp(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6),
  `deleted` tinyint(1) DEFAULT '0',
  PRIMARY KEY (`message_id`),
  UNIQUE KEY `message_id_UNIQUE` (`message_id`)
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8;
```

Debug output:

```
Add command: prepare
0 5828 <== prepare#start(0,,142)
0 5828 ==> prepare#prepareHeader(1,maybeOK,12)
 raw: 000300000005000300000000
0 5828 ==> prepare#readParameter(2,,23)
 raw: 03646566000000013f000c3f0000000000fd8000000000
0 5828 ==> prepare#readParameter(3,,23)
 raw: 03646566000000013f000c3f0000000000fd8000000000
0 5828 ==> prepare#readParameter(4,,23)
 raw: 03646566000000013f000c3f0000000000fd8000000000
0 5828 ==> prepare#parametersEOF(5,EOF,5)
 raw: fe00000200
0 5828 ==> prepare#readField(6,,57)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d657373616765730466726f6d0466726f6d0c21002d000000fd0110000000
0 5828 ==> prepare#readField(7,,67)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d657373616765730974696d657374616d700974696d657374616d700c3f001a000000078124060000
0 5828 ==> prepare#readField(8,,57)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d6573736167657304626f647904626f64790c2100ee020000fd0110000000
0 5828 ==> prepare#readField(9,,63)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d6573736167657307757365725f696407757365725f69640c3f000b000000030110000000
0 5828 ==> prepare#readField(10,,69)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d657373616765730a6d6573736167655f69640a6d6573736167655f69640c3f000b000000030742000000
0 5828 ==> prepare#fieldsEOF(11,EOF,5)
 raw: fe00000200
Add command: executeStatement
0 5828 <== executeStatement#start(0,,41)
0 5828 ==> executeStatement#resultsetHeader(1,,1)
 raw: 05
        Resultset header received, expecting 5 column definition packets
0 5828 ==> executeStatement#readField(2,,57)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d657373616765730466726f6d0466726f6d0c21002d000000fd0110000000
0 5828 ==> executeStatement#readField(3,,67)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d657373616765730974696d657374616d700974696d657374616d700c3f001a000000078124060000
0 5828 ==> executeStatement#readField(4,,57)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d6573736167657304626f647904626f64790c2100ee020000fd0110000000
0 5828 ==> executeStatement#readField(5,,63)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d6573736167657307757365725f696407757365725f69640c3f000b000000030110000000
0 5828 ==> executeStatement#readField(6,,69)
 raw: 036465660b65726963765f70686f6e65086d65737361676573086d657373616765730a6d6573736167655f69640a6d6573736167655f69640c3f000b000000030742000000
0 5828 ==> executeStatement#fieldsEOF(7,EOF,5)
 raw: fe00002200
Compiled binary protocol row parser:
(function(){ return function BinaryRow(packet) {
  var statusByte = packet.readInt8();
  var nullBitmaskByte0 = packet.readInt8();
  // "from": VAR_STRING
  if (nullBitmaskByte0 & 4)
    this["from"] = null;
  else
    this["from"] = packet.readLengthCodedString();
  // "timestamp": TIMESTAMP
  if (nullBitmaskByte0 & 8)
    this["timestamp"] = null;
  else
    this["timestamp"] = packet.readDateTime();
  // "body": VAR_STRING
  if (nullBitmaskByte0 & 16)
    this["body"] = null;
  else
    this["body"] = packet.readLengthCodedString();
  // "user_id": LONG
  if (nullBitmaskByte0 & 32)
    this["user_id"] = null;
  else
    this["user_id"] = packet.readSInt32();
  // "message_id": LONG
  if (nullBitmaskByte0 & 64)
    this["message_id"] = null;
  else
    this["message_id"] = packet.readSInt32();
}; })()
0 5828 ==> executeStatement#row(8,maybeOK,39)
 raw: 00000d2b3434373531383638323338390be007011510230d670a09000248690100000001000000
0 5828 ==> executeStatement#row(9,maybeOK,41)
 raw: 00000d2b3434373531383638323338390be007011510242b6a9f080004596f75210100000002000000
0 5828 ==> executeStatement#row(10,maybeOK,52)
 raw: 00000d2b3434373531383638323338390be007011510242fe01c0a000f486f77277320697420676f696e673f0100000003000000
```

Cleaned statement simplified:
```
var query = "SELECT `from`, `timestamp`, `body`, `user_id`, `message_id` FROM `messages` WHERE `user_id` = :user_id AND `from` = :from AND `deleted` = 0 LIMIT :num;";

connection.execute(query, {user_id:1, from:1, num:1}, function(err, rows){
            if(err){
                console.log(err);
                connection.release();
                callback(err);
            }else{
                connection.release();
                callback(rows);
            }
        });

```
Insert statement for test data:

```
INSERT INTO `messages` (`from`, `body`, `user_id`, `message_id`) VALUES (1, "bleh", 1, 1);
```

**Comment by sidorares** [comment]
Thanks, I'll try to reproduce

**Comment by ericvolp12** [comment]
Great, thanks. Good luck, this has been a pain but it seems like such a weird issue to stall me for so long :(

**Comment by ericvolp12** [comment]
Any luck?

**Comment by sidorares** [comment]
```
mysql> INSERT INTO `messages` (`from`, `body`, `user_id`, `message_id`) VALUES (1, "bleh", 1, 1);
ERROR 1364 (HY000): Field 'to' doesn't have a default value
```

**Comment by ericvolp12** [comment]
Oops, make to '1' as well

**Comment by sidorares** [comment]
ok, have error as well, great

**Comment by ericvolp12** [comment]
Okay, does that mean it was replicatable or did something else go wrong?

**Comment by sidorares** [comment]
I meant able to replicate same error as you having
Might be a bug in some type de-serializing. In the meantime until I find the problem - `connection.query` worked for me with same data / query

**Comment by ericvolp12** [comment]
Ah, okay. Does that do the same prepared statement stuff? I use execute for all my other queries so I'll just have to build query into my db module.

**Comment by sidorares** [comment]
so the bug was in this line - https://github.com/sidorares/node-mysql2/blob/baab1c6dc8f6dd3e51a3aa72987ce7cde52f8051/lib/packets/packet.js#L187

should be `>= 11` ( or `> 10` )

spec is here - http://dev.mysql.com/doc/internals/en/binary-protocol-value.html#packet-ProtocolBinary::MYSQL_TYPE_DATE

could you edit your `/lib/packets/packet.js` file and try?

Looks like long 11-byte version of ProtocolBinary::MYSQL_TYPE_DATE packet was never tested

**Comment by sidorares** [comment]
> Does that do the same prepared statement stuff

no, `query()` does not use prepared statements and response is serialized using much simpler text protocol ( basically array of length-encoded strings for each row ) - http://dev.mysql.com/doc/internals/en/com-query-response.html#packet-ProtocolText::ResultsetRow

**Comment by ericvolp12** [comment]
Awesome, that fixed it, thanks!

**Comment by sidorares** [comment]
I'll keep it open until I push fix + test

**Comment by ericvolp12** [comment]
Okay, sorry :P

**Comment by sidorares** [comment]
@ericvolp12 published as v1.0.0-rc.1
